### PR TITLE
Move network information from `Identity` to `IdentityContext`

### LINKED ISSUE #1841: Move identity network type to IdentityContext
### Context

Many blockchains "share" the same key pair construction mechanism, which means owning one private key would grant you access to the same public key or address on different networks.

Examples:
- EVM based: ETH, Polygon, BSC -> they have the same public key and address
- Substrate based: relaychains + parachains -> they have the same public key, but different addresses due to different ss58 prefixes
 
We [had discussions](https://litentry.slack.com/archives/C03KR96989G/p1680718905532839?thread_ts=1680708645.416489&cid=C03KR96989G) about whether we should treat the same public key on different networks as distinct identities, the decision was _yes_ at that time. Reasons include:
- it gives the user finer control over their identities, e.g. they can link their polkadot address to the IDGraph and intentionally not link their kusama address
- without explicit network types, calculating specific VC would be unrealistic, e.g. the number of governance activities over all linked substrate identities

During the implementation of EVM-signin feature, we realised that we might be able to simplify this process - by moving the network type information away from `Identity` itself to `IdentityContext`.

**Before:**

```
pub enum Identity {
	Substrate { network: SubstrateNetwork, address: Address32 },
	Evm { network: EvmNetwork, address: Address20 },
	Web2 { network: Web2Network, address: IdentityString },
}

pub struct IdentityContext<T: Config> {
	// the sidechain block number at which the identity is linked
	pub link_block: BlockNumberOf<T>,
	// the identity status
	pub status: IdentityStatus,
}
```

**After:**
```
pub enum Identity {
        // web2
        Twitter(IdentityString),
        Discord(IdentityString),
        Github(IdentityString),
        // web3
	Substrate(Address32),
	Evm(Address20),
	
}

pub struct IdentityContext<T: Config> {
	// the sidechain block number at which the identity is linked
	pub link_block: BlockNumberOf<T>,
        // a list of networks on which this identity is used
        pub networks: Vec<Network>, 
	// the identity status
	pub status: IdentityStatus,
}
```

**Rationale:**
- **Foremost**: the TrustedCall/DI is authorised using signature, and **signature is ignorant of network types**. It's impossible to specify the identity network that can access/manage the IDGraph. So you can't say: _we only allow litentry identities to manage the IDGraph, not polkadot identities_. It doesn't make sense because the signature doesn't contain the network information, they are just calculated from key pairs
- it makes the linking process easier for the user. They don't have to repeatedly link identities with different networks despite sharing the same public key (thus the same validation data). They only link it once.
- it reduces the IDGraph size.
- it _might_ lower the chance of cheating, given that one identity can only exist in one IDGraph and you can't unlink it.

### Proposal

Meanwhile, we want to maintain flexibility, and the user should still have complete control over things like identities on which network should take part in the assertion building.

While we don't distinguish network types when adding identities to IDGraph, sometime, somewhere, we need to ask the user:  for which networks do you want your identities in IDGraph to be used? It can be 3 such moments:
1. when the user links an identity
2. when the user builds a VC 
3. somewhere in between

We propose a hybrid solution of 1 and 3.

Example:
- when the user links a substrate identity (EVM-based is similar), we give the user the initial possibility to specify which networks the identity should be used for. In UI it's something like a multi-select box, where we have some default networks that are ticked already (e.g. Litentry, polkadot, kusama, phala...) but the user is free to check or uncheck the network. Just so you know, it's a finite set of networks, mostly dependent on the data provider, so it won't be 100 parachain networks.
- we provide an extrinsic or DI called `set_identity_network` so that user can update the network list at any time
- the change will take effect in the next VC generation. So for example, when the user only sets `kusama` as active network, only activities on `kusama` will be considered when the user requests the VC "_number of governance activities over all linked substrate identities_", no polkadot, no litentry or any other substrate networks.

---
:heavy_check_mark: Please set appropriate **labels** and **assignees** if applicable.

### PR DESCRIPTION
**Pull request by Kailai-Wang**
### Context

Resolves #1841 
It's a **breaking change** for the client.

This PR tries to:
- relocate the network information from the `Identity` struct to `IdentityContext` struct, it uses `Vec<Web3Network>` to highlight that it's only related to web3 identities
- restructure the `Identity` and `Web3Network` to be more flattened, see see `network.rs` - **please let me know the feedback** (tagging @jonalvarezz on this too)
- replace `SupportedNetwork` and `IndexingNetwork` with `Web3Network` in assertion building
- specify the supported networks for each `Assertion`: `get_supported_web3networks`
- simplify the assertion building logic
- unify the logic of `A4/A7/A10/A11`, but probably needs further refactoring, see #1813 
- remove the deprecated `mock-tee-primitives` package
- fix a validation data signing bug in ts-test

TODO:
- add an interface to allow users to update the web3networks, will follow this up once this PR is merged (because I want to make sure the fundamentals are aligned before moving on)
- maybe more testing regarding the networks in ts-test

### CODE REVIEW

File: pallets/identity-management/src/lib.rs
Comment (BillyWooo) [inline pallets/identity-management/src/lib.rs:223]: why it's not used?

File: tee-worker/cli/src/base_cli/commands/litentry/link_identity.rs
Comment (BillyWooo) [inline tee-worker/cli/src/base_cli/commands/litentry/link_identity.rs:69]: which params are incorrect?

File: pallets/identity-management/src/lib.rs
Comment (Kailai-Wang) [reply pallets/identity-management/src/lib.rs:223]: It's used in indirect calls

File: tee-worker/cli/src/base_cli/commands/litentry/link_identity.rs
Comment (Kailai-Wang) [reply tee-worker/cli/src/base_cli/commands/litentry/link_identity.rs:69]: The extrinsic parameters - they are a bit outdated

File: primitives/core/src/network.rs
Comment (kziemianek) [inline primitives/core/src/network.rs:50]: I'm not sure if this is good structure. For example what if we want to add new substrate network, would it be added at the end of 'substrate' or after 'evm' values ? I think one way might be breaking change as this is enumerated type.

Have you considered something like this?
```
pub enum Web3Network {
	Substrate(SubstrateNetwork),
	Evm(EvmNetwork)
}

pub enum SubstrateNetwork {
	Polkadot,
	Kusama,
	Litentry,
	Litmus,
	LitentryRococo,
	Khala,
	SubstrateTestnet, // when launched it with standalone (integritee-)node
}

pub enum EvmNetwork {
	Ethereum,
	Polygon,
	BSC,
}
```

This would also reduce the need for `is_substrate` and `is_evm` methods.

File: primitives/core/src/network.rs
Comment (kziemianek) [inline primitives/core/src/network.rs:67]: Have you considered using the normal match as it's exhaustive and requires you handling a new variant. Adding following test could also prevent from such mistakes:

```
	#[test]
	fn is_evm_network() {
		Web3Network::iter().for_each(|network| {

			assert_eq!(network.is_evm(), match network {
				Web3Network::Polkadot => false,
				Web3Network::Kusama => false,
				Web3Network::Litentry => false,
				Web3Network::Litmus => false,
				Web3Network::LitentryRococo => false,
				Web3Network::Khala => false,
				Web3Network::SubstrateTestnet => false,
				Web3Network::Ethereum => true,
				Web3Network::Polygon => true,
				Web3Network::BSC => true
			})
		})
	}
```

File: tee-worker/litentry/primitives/src/identity.rs
Comment (kziemianek) [inline tee-worker/litentry/primitives/src/identity.rs:71]: Potentialy same problem as with `Web3Network`

**Review by kziemianek — COMMENTED** [review]
We are very often leaving todos in code, do we have any strategy for resolving them 🙃  ?

File: primitives/core/src/network.rs
Comment (BillyWooo) [reply primitives/core/src/network.rs:50]: It was something like that before the change. One more structure layer in between. Either way has its stand. I think here Kai wants to decrease the layers.

Comment (Kailai-Wang) [reply primitives/core/src/network.rs:50]: Hi @kziemianek, yes I intentionally did so after pondering it 🥲 see the comment just above this struct:
```
/// supported web3 networks
/// use a flattened style to avoid overly nested structure like:
/// {
///   Substrate(SubstrateNetwork),
///   Evm(EvmNetwork),
/// }
```

The nested style would complicate the construction of such type (also from F/E side). Compare this: 
```
Web3Network::Litentry
```
vs
```
Web3Network::Substrate(SubstrateNetwork::Litentry)
```
Both "Substrate" are kind of redundant, but you need to have some names for the enum types, which is a bit awkward. You can have extra `From` or `TryFrom` implementation but that's not as straightforward either.

The value retrieval and comparison/branching typically need more effort (`if let`/`match`) in code which lowers the readability. 

If some of the enum types (e.g. `Ethereum`) need further subdivision, that's even more layers. `is_substrate` and `is_evm` would take a one-time maintenance effort so I'm fine with it.

Adding new values is a fair point, but either way it would need a runtime upgrade to take effect => it will have a new spec_version number => it signals the client it's a breaking change => the client needs to refresh/renew the metadata eventually. That said I'll keep it categorised when adding new values.

So in general I consider the flattened structure a bit more advantageous but I'm open to suggestions.

File: primitives/core/src/network.rs
Comment (Kailai-Wang) [reply primitives/core/src/network.rs:67]: That's a good point - I'll add test cases to prevent errors due to being silently ignored

**Comment by Kailai-Wang** [comment]
> We are very often leaving todos in code, do we have any strategy for resolving them 🙃 ?

In general we should follow up with issues if it's not so trival.

I normally leave `TODO`s since:
- it's better/clearer for a separate PR
- it's not a trivial effort
- I accidentally find something that shouldn't belong to this PR, to not forget about it
- I know it can be improved, but have no idea right now

Maybe I've left too many `TODO`s - my apology.

File: tee-worker/litentry/primitives/src/identity.rs
Comment (Kailai-Wang) [reply tee-worker/litentry/primitives/src/identity.rs:71]: Yes same consideration as above

File: primitives/core/src/network.rs
Comment (kziemianek) [reply primitives/core/src/network.rs:50]: _Adding new values is a fair point, but either way it would need a runtime upgrade to take effect => it will have a new spec_version number => it signals the client it's a breaking change => the client needs to refresh/renew the metadata eventually. That said I'll keep it categorised when adding new values._

This might not be true for clients that are using DI only.

This scaries me a bit more, because if you didn't notice the change, you might be sending old values and everything will work fine, except the results will be not the ones you expected.

Comment (kziemianek) [reply primitives/core/src/network.rs:50]: Keeping this structure but not forcing new variants to be added "in between" should avoid such problems.

Comment (Kailai-Wang) [reply primitives/core/src/network.rs:50]: That's akin to the substrate extrinsic index - where adding new extrinsic is backwards compatible, but if you change the order that's a breaking change and you have to bump the `spec_version`/`transaction_version`.

I'd argue even if you split into `evm` and `substrate` groups you can still fall into this problem once you remove/change the existing fields. The same applies to other enums too (remember the famous `trusted_call`).

I'm not sure how IDHub maintains the generated types, but maybe we can minimise the risk by having a proper version and metadata update mechanism, like utilise the runtime version in sidechain too - once the client sees a different runtime version, it should realise something is broken and they should re-fetch the metadata and re-generate the types.

File: tee-worker/litentry/pallets/identity-management/src/identity_context.rs
Comment (kziemianek) [inline tee-worker/litentry/pallets/identity-management/src/identity_context.rs]: What do you think about removing duplicates from `web3networks` here ?

File: tee-worker/litentry/pallets/identity-management/src/lib.rs
Comment (kziemianek) [inline tee-worker/litentry/pallets/identity-management/src/lib.rs:141]: The more I think about it the more I'm convinced that we should let user decide when setting shielding key.

Apart from that, does it make sense to set "evm" networks for `Identity::Substrate` ?

File: tee-worker/litentry/pallets/identity-management/src/lib.rs
Comment (kziemianek) [inline tee-worker/litentry/pallets/identity-management/src/lib.rs:181]: I think we should make sure that linked identity and networks belong to same type of chain ("substrate"/"evm").

Having such mix doesn't make sense and should be prevented from happening.

If applied, it would be good to have a test for it.

File: tee-worker/litentry/primitives/src/identity.rs
Comment (kziemianek) [inline tee-worker/litentry/primitives/src/identity.rs:86]: Might be worth to have same tests as for `Web3Network` 🙃?

File: tee-worker/litentry/core/stf-task/sender/src/lib.rs
Comment (kziemianek) [inline tee-worker/litentry/core/stf-task/sender/src/lib.rs]: What do you think about promoting it to the new type, for example 'IdentityWithNetworks` ?
This tuple is widely used across assertions, which makes code less readable.

File: tee-worker/litentry/core/assertion-build/src/a1.rs
Comment (kziemianek) [inline tee-worker/litentry/core/assertion-build/src/a1.rs:31]: We don't want to have such sensitive information to appear in logs or is it jus irrevelant ?

File: tee-worker/litentry/core/assertion-build/src/a7.rs
Comment (kziemianek) [inline tee-worker/litentry/core/assertion-build/src/a7.rs:42]: I think we should filter out all "identities" having `Web3Network` other than Polkadot as it doesn't make sense to ask for them in context of this assertion or it may lead to false positive results.

File: tee-worker/litentry/core/assertion-build/src/a8.rs
Comment (kziemianek) [inline tee-worker/litentry/core/assertion-build/src/a8.rs:86]: One day it may panic 😂

**Review by kziemianek — COMMENTED** [review]
I gave it a second more careful look.

File: tee-worker/litentry/core/assertion-build/src/a7.rs
Comment (Kailai-Wang) [reply tee-worker/litentry/core/assertion-build/src/a7.rs:42]: Maybe this is what you meant? https://github.com/litentry/litentry-parachain/blob/08a7b9201035753dc8d344b69bc09e80ac3c25be/tee-worker/app-libs/stf/src/trusted_call_litentry.rs#L141-L148

File: tee-worker/litentry/core/assertion-build/src/a8.rs
Comment (Kailai-Wang) [reply tee-worker/litentry/core/assertion-build/src/a8.rs:86]: Noooo - then rust is broken.

The best thing I can do is use default value if `try_into` fails 😂 
If we fall into this branch, it means something is wrong when generating VC and we need to report it back to parachain, so we can't have any other error types than `RequestVCFailed`

File: tee-worker/litentry/pallets/identity-management/src/identity_context.rs
Comment (Kailai-Wang) [reply tee-worker/litentry/pallets/identity-management/src/identity_context.rs]: Do you mean the client could insert duplicate types in this vector?

File: tee-worker/litentry/primitives/src/identity.rs
Comment (Kailai-Wang) [reply tee-worker/litentry/primitives/src/identity.rs:86]: Good call

File: tee-worker/litentry/pallets/identity-management/src/lib.rs
Comment (BillyWooo) [reply tee-worker/litentry/pallets/identity-management/src/lib.rs:141]: I would suggest not to do so. If we let the user to decide when to set shielding key, it's out of control. The user may not set any shielding key. 
What do you mean by **set "evm" networks for `Identity::Substrate`** ?

Comment (Kailai-Wang) [reply tee-worker/litentry/pallets/identity-management/src/lib.rs:141]: I agree it needs to be discussed. To me it's more like a product/user experience issue, so I look for opinions of @kaylawangnow @jonalvarezz @mikuczi

One opposing view would be: when the user sets the shielding key, it's confusing that they need to select the identity type - at least less natural than linking an identity. If they want to change the network information of the prime account, they can use the `set_identity_web3network` at any time

File: tee-worker/litentry/pallets/identity-management/src/identity_context.rs
Comment (kziemianek) [reply tee-worker/litentry/pallets/identity-management/src/identity_context.rs]: Yes, but whatever the reason is, we want to prevent from having same network enabled multiple times.

File: tee-worker/litentry/core/assertion-build/src/a7.rs
Comment (kziemianek) [reply tee-worker/litentry/core/assertion-build/src/a7.rs:42]: Yes, but the purpose of this code is to reduce request size, not to prevent assertions from making wrong requests.
I think it could be worth to add redundant check. Better safe than sorry.

File: tee-worker/litentry/pallets/identity-management/src/lib.rs
Comment (kziemianek) [reply tee-worker/litentry/pallets/identity-management/src/lib.rs:141]: > I would suggest not to do so. If we let the user to decide when to set shielding key, it's out of control. The user may not set any shielding key.

@BillyWooo What I mean is that the setting user shielding key use cause should invole picking networks for prime identity by user explicitly.

> What do you mean by set "evm" networks for Identity::Substrate ?

@BillyWooo This code will result in substrate identity being linked to user's IdGraph with all web3 networks activated, even Ethereum, Polygon and BSC. Will it be fine?

```
let web3networks = get_all_web3networks()
    .try_into()
    .map_err(|_| Error::<T>::Web3NetworkLenLimitReached)?;
let context = <IdentityContext<T>>::new(
    <frame_system::Pallet<T>>::block_number(),
    web3networks,
);
Self::insert_identity_with_limit(&who, &prime_id, context)?;
```

Comment (Kailai-Wang) [reply tee-worker/litentry/pallets/identity-management/src/lib.rs:141]: Currently people link identities via IDHub, where it comes down to two distinct categories: substrate or evm. So as long as F/E send the right types it's all good.

But yes from code logic we should have a sanity check here. A follow-up question is how to handle it? Do you simply throw out an error, or silently leave out the unmatched network types? It may be another user experience problem @kaylawangnow @jonalvarezz

File: tee-worker/litentry/core/assertion-build/src/a7.rs
Comment (Kailai-Wang) [reply tee-worker/litentry/core/assertion-build/src/a7.rs:42]: I don't get it - the `assertion_networks` already contains the desired network, doesn't it? E.g. for `A7` it would be `Web3Network::Polkadot`

File: tee-worker/litentry/core/stf-task/sender/src/lib.rs
Comment (Kailai-Wang) [reply tee-worker/litentry/core/stf-task/sender/src/lib.rs]: Will do :)

File: tee-worker/litentry/core/assertion-build/src/a7.rs
Comment (kziemianek) [reply tee-worker/litentry/core/assertion-build/src/a7.rs:42]: Yes, the code in current state should work fine.

File: tee-worker/litentry/core/assertion-build/src/a6.rs
Comment (kziemianek) [inline tee-worker/litentry/core/assertion-build/src/a6.rs:38]: We don't want to have such sensitive information to appear in logs or is it jus irrevelant ?

File: tee-worker/litentry/core/assertion-build/src/a1.rs
Comment (kziemianek) [reply tee-worker/litentry/core/assertion-build/src/a1.rs:31]: This should point to other assertion 🙃 
https://github.com/litentry/litentry-parachain/pull/1854/files/08a7b9201035753dc8d344b69bc09e80ac3c25be#r1255746447

Comment (Kailai-Wang) [reply tee-worker/litentry/core/assertion-build/src/a1.rs:31]: Irrelevant I think - you will see who has requested a VC (via extrinsics) - who has received a VC (via `VCIssued` event) - so it should be fine here

File: tee-worker/litentry/pallets/identity-management/src/lib.rs
Comment (Kailai-Wang) [reply tee-worker/litentry/pallets/identity-management/src/lib.rs:141]: I currently throw out an error if the types don't match, because 99% of users (exception: IDHub team) will link it using IDHub so I don't expect it to go wrong. If it however does, it must be a bug

File: primitives/core/src/network.rs
Comment (jonalvarezz) [reply primitives/core/src/network.rs:50]: I also agree that we should look for structures that keep breaking changes to a minimum or are at least backward compatible in the future without translating the problem to clients.

put another way, as a client, if you give to pick between ugly and nested structures that require mininum to non maintenance and nice structures that require me to watch every single thing you change, I'd very much prefer the former.

Now, I leave it to you folks to find such a structure and keep a good balance between both trade-offs 👍

**Review by kziemianek — APPROVED** [review]
Looks good to me.
It's great that week ago the idea came up and we already have an implementation. Nice pace 🚀

File: tee-worker/litentry/pallets/identity-management/src/lib.rs
Comment (jonalvarezz) [reply tee-worker/litentry/pallets/identity-management/src/lib.rs:141]: > The more I think about it the more I'm convinced that we should let user decide when setting shielding key

This is a very fair point. My reasoning from a user perspective is that since:
1. The user can update their linked networks at anytime, including disabling previously activated networks, and
2. This happens only for the main (prime) account

It is alright. 

I'd suggest to however leave this to clients, i.e., don't do any automatic network activation here but accept a `web3network` vec with the list of networks to activate as part of setting the shielding key, if any. 

This enable clients to do it under the hood or to display the option to the user. IDHub, as requested by Product, will do it automatically at the beginning but it leaves the door open to change it at anytime.

Additionally, another reason to not perform any automatic activation here is that we are using setUserShieldingKey as a recovery method, e.g., you forgot your password. It wouldn't be nice to get networks activating as a side effect.

Kasper: If the IDHub goes with this automatic network linking when setting shielding key, we will make sure to let the user know what will happen so they know it upfront.

Comment (jonalvarezz) [reply tee-worker/litentry/pallets/identity-management/src/lib.rs:141]: > But yes from code logic we should have a sanity check here. A follow-up question is how to handle it? Do you simply throw out an error, or silently leave out the unmatched network types? 

Throwing an error if any of networks doesn't match sounds totally sane. A partial success would be more complex to properly handle. Make it fail out loud 

> I currently throw out an error if the types don't match, because 99% of users (exception: IDHub team) will link it using IDHub so I don't expect it to go wrong. If it however does, it must be a bug

Do we have the data to deny or confirm that IDHub is the only client out there using our Parachain? It is public, anyone could be using it and toying around with it. as far as I know, we are not monitoring it extensively either 😄

Futhermore, I don't think it is healthy for us, as a public parachain, to think only in terms of IDHub. We should always use the word "clients" and tackle issues holistically as service provider. This is, to not minimize the impact of issues because we can translate and control them within the IDHub.

Comment (Kailai-Wang) [reply tee-worker/litentry/pallets/identity-management/src/lib.rs:141]: > Do we have the data to deny or confirm that IDHub is the only client out there using our Parachain? 

No - it's completely open.
But if not using IDHub, probably only tech guys can construct the parameters correctly by looking up the code, if they are that far, they should know the handling of mismatching networks in our code.

**Comment by Kailai-Wang** [comment]
Please don't merge it now - we are still having discussions about the struct/enum :(

**Review by BillyWooo — APPROVED** [review]
LGTM. Excellent. I think it's good to go.

**Comment by Kailai-Wang** [comment]
We had some discussions about whether we should go back to the old structure and change the network to a list of vectors, like this:
```
pub enum Identity {
	Substrate(Address32, BoundedVec<SubstrateNetwork, ..>),
	Evm(Address20, BoundedVec<EvmNetwork, ...>),
	Web2(IdentityString, Web2Network),
}
```

but it will be problematic if we need to pass in those identities as parameters in e.g. `remove_identity` because we need to correctly construct those network types.

**Pull request merged**
